src/main/scala/magellan/Polygon.scala

```diff
@@ -158,7 +158,15 @@ class Polygon extends Shape {
     * @return
     */
   private [magellan] def intersects(polygon: Polygon): Boolean = {
-    polygon.loops.exists(otherLoop => this.loops.exists(_.intersects(otherLoop)))
+    var intersects = false
+    if(polygon.getVertexes().exists(other => this.contains(other))
```

Comment:
is there a more efficient solution here? every non degenerate use case will have to run two expensive polygon contains point checks before checking line intersection

Reply:
Well, short answer. There is a more efficient solution. (As always!!!)
I believe we can have really fast algorithms looking first at the indexes that have been created. 
Eg. I big polygon with inner extent that can be represented as  geohashing with 4 length string can be easily be compared with a point(s) when there are geohashed. (That's really fast!!!!!)

I also have in mind this lib (http://www.angusj.com/delphi/clipper.php ), which claims that is the fastest of all.

Reply:
I don't think we need to look st third party algorithm
What if we did this?
The only time you have a polygon within another is when their bounding boxes are within each other... why don't we check for that edge case before we do the contains?

Reply:
well, you can't check the extent for intersection, just for a hint that the polygons are close to each other.  
what was thinking was to see how the clipper library is doing the intersections and then implement it to magellan

Reply:
Ok. Can you look into the clipper library and revert back with what we need in Magellan to do this efficiently? 
Indexes can be handled separately, just look to make the non index codepath efficient enough that it doesn't do wasteful computation in generic intersections just to account for edge cases like containment

Reply:
on second thought, why doesn't this work?
a) check intersects first, i.e. if any of the edges of this polygon intersects the other. if so we are done. (takes care of generic intersections)
b) if not, test if the bounding box of one shape is fully within the other. 
    if false, return false
   else check containment

Reply:
That does catch all the case. I will try to draw few examples with polygons, so I can make it clear.

Reply:
![not_intersect](https://user-images.githubusercontent.com/8981766/31276310-e1f8b4c0-aaa3-11e7-83d9-ea5cfa7bdf08.jpg)
1. The polygons edges don't intersect and extents are inside the other

Reply:
![not_intersect2](https://user-images.githubusercontent.com/8981766/31276382-39b49008-aaa4-11e7-8430-9273fb53d307.jpg)
2. Polygon with holes, not intersecting with the start. The edges don't intersect but the extents are inside the other.

Reply:
In the second step if the two boundary boxes are intersecting, then check containment, 
else return false.

Reply:
And then if you check the intersection of the extents first and it is false, then you don't have to check the edges intersection.

Reply:
btw it might be worth spending some time researching algorithms here... let's circle back in a couple of days with some options

Reply:
I agree